sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+        )
+    }
+
+    /// Swap `Coin<T>` for the `Coin<SUI>`.
+    public(script) fun swap_token<T>(
+        pool: &mut Pool<T>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        let tok_balance = Coin::into_balance(token);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            Balance::value(&tok_balance),
+            token_reserve,
+            sui_reserve,
+        );
+
+        Balance::join(&mut pool.token, tok_balance);
+        Transfer::transfer(
+            Coin::withdraw(&mut pool.sui, output_amount, ctx),
+            TxContext::sender(ctx)
+        )
+    }
+
+    /// Add liquidity to the `Pool`. Sender needs to provide both
+    /// `Coin<SUI>` and `Coin<T>`, and in exchange he gets `Coin<LSP>` -
+    /// liquidity provider tokens.
+    public(script) fun add_liquidity<T>(
+        pool: &mut Pool<T>, sui: Coin<SUI>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        let sui_balance = Coin::into_balance(sui);
+        let token_balance = Coin::into_balance(token);
+
+        let (sui_amount, _, lsp_supply) = get_amounts(pool);
+
+        let sui_added = Balance::value(&sui_balance);
+        let share_minted = (sui_added * lsp_supply) / sui_amount;
+
+        // TODO: Figure out what to do with this buddy.
+        // let coin2_added = (share_minted * tok_amount) / lsp_supply;
+
+        Balance::join(&mut pool.sui, sui_balance);
+        Balance::join(&mut pool.token, token_balance);
+
+        Transfer::transfer(
+            Coin::mint(share_minted, &mut pool.lsp_treasury, ctx),
+            TxContext::sender(ctx)
+        );
+    }
+
+    /// Remove liquidity from the `Pool` by burning `Coin<LSP>`.
+    /// Sender gets `Coin<T>` and `Coin<SUI>`.
+    public(script) fun remove_liquidity<T>(
+        pool: &mut Pool<T>,
+        lsp: Coin<LSP<T>>,
+        ctx: &mut TxContext
+    ) {
+        let lsp_amount = Coin::burn(lsp, &mut pool.lsp_treasury);
+        let (sui_amt, tok_amt, lsp_supply) = get_amounts(pool);
+
+        let sui_removed = (sui_amt * lsp_amount) / lsp_supply;
+        let tok_removed = (tok_amt * lsp_amount) / lsp_supply;
+
+        let sender = TxContext::sender(ctx);
+
+        Transfer::transfer(
+            Coin::withdraw(&mut pool.sui, sui_removed, ctx),
+            sender
+        );
+
+        Transfer::transfer(
+            Coin::withdraw(&mut pool.token, tok_removed, ctx),
+            sender
+        );
+    }
+
+    /// Get most used values in a handy way:
+    /// - amount of SUI
+    /// - amount of token
+    /// - total supply of LSP
+    public fun get_amounts<T>(pool: &Pool<T>): (u64, u64, u64) {
+        (
+            Balance::value(&pool.sui),
+            Balance::value(&pool.token),
+            Coin::total_supply(&pool.lsp_treasury)
+        )
+    }
+
+    /// Calculate the output amount minus the fee - 0.3%
+    fun get_input_price(
```

Comment:
I think it is also useful to expose public `get_sui_price` and `get_t_price` functions--on-chain prices w/o oracles is one of the very cool things about AMM's.